File: TestServer/connect2.py
import socket
import struct

def connect_to_server(host, port):
    client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    client_socket.connect((host, port))
    return client_socket

def send_initial_request(client_socket):
    client_socket.send(b'\x01')

def receive_snapshot_data(client_socket):
    # Read the packet length field
    packet_length_data = client_socket.recv(4)
    packet_length = struct.unpack('<i', packet_length_data)[0]
    print(packet_length)

    # Read the remaining packet data
    packet_data = client_socket.recv(126)

    # Parse and process the packet data
    print(packet_data[:30])
    symbol = packet_data[:30].decode('utf-8').strip('\x00')
    sequence_number = struct.unpack('<q', packet_data[30:38])[0]
    timestamp = struct.unpack('<q', packet_data[38:46])[0]
    ltp = struct.unpack('<q', packet_data[46:54])[0] / 100
    ltp_quantity = struct.unpack('<q', packet_data[54:62])[0]
    volume = struct.unpack('<q', packet_data[62:70])[0]
    bid_price = struct.unpack('<q', packet_data[70:78])[0] / 100
    bid_quantity = struct.unpack('<q', packet_data[78:86])[0]
    ask_price = struct.unpack('<q', packet_data[86:94])[0] / 100
    ask_quantity = struct.unpack('<q', packet_data[94:102])[0]
    open_interest = struct.unpack('<q', packet_data[102:110])[0]
    prev_close_price = struct.unpack('<q', packet_data[110:118])[0] / 100
    prev_open_interest = struct.unpack('<q', packet_data[118:126])[0]

    # Process the received snapshot data
    print("Symbol:", symbol)
    print("Sequence Number:", sequence_number)
    print("Timestamp:", timestamp)
    print("Last Traded Price (LTP):", ltp)
    print("Last Traded Quantity:", ltp_quantity)
    print("Volume:", volume)
    print("Bid Price:", bid_price)
    print("Bid Quantity:", bid_quantity)
    print("Ask Price:", ask_price)
    print("Ask Quantity:", ask_quantity)
    print("Open Interest (OI):", open_interest)
    print("Previous Close Price:", prev_close_price)
    print("Previous Open Interest:", prev_open_interest)
    print("---------------------")

# Set the host and port based on the server configuration
host = 'localhost'
port = 9011

# Connect to the server
client_socket = connect_to_server(host, port)

# Send the initial request
send_initial_request(client_socket)

# Continuously receive and process the snapshot data
while True:
    try:
        receive_snapshot_data(client_socket)
    except KeyboardInterrupt:
        break
    # except:
    #     pass

# Close the connection
client_socket.close()

File: TestServer/connect2.js
const app = require("express")();
const http = require("http").Server(app);
const io = require("socket.io")(http, {
    cors: {
        origin: '*',
    }
});;
const net = require('net');

// Port Config
const host = 'localhost';
const port = 9011;

io.on('connection', (socket) => {
    console.log('Socket client connected');
    // Handle snapshot data requests from the client
    socket.on('requestSnapshotData', () => {
        connectToServer(socket);
    });

    // Handle disconnection
    socket.on('disconnect', () => {
        console.log('Socket client disconnected');
    });
});

function connectToServer(socket) {
    const client = net.createConnection({ host, port }, () => {
        console.log('Connected to Market Data Streaming Server...');
        console.log('Sending Initial Byte Stream....')
        // Send the initial request
        client.write(Buffer.from([0x01])); // Send a single byte with value 0x01
    });

    let receivedData = Buffer.alloc(0);

    client.on('data', (data) => {
        receivedData = Buffer.concat([receivedData, data]);

        while (receivedData.length >= 130) {
            const packetData = receivedData.slice(0, 130);
            receivedData = receivedData.slice(130);

            receiveSnapshotData(packetData, socket);
        }
    });

    client.on('close', () => {
        console.log('Connection closed');
    });

    client.on('error', (error) => {
        console.error('An error occurred:', error.message);
    });
}

function receiveSnapshotData(data, socket) {
    // Parse and process the packet data

    const symbol = data.toString('utf-8', 4, 34).replace(/\0+$/, '');
    var underlying = "";
    var expiryDate = "";
    var strikePrice = "";
    var optionType = "";

    let endIndex = -1;
    ['ALLBANKS', 'MAINIDX', 'FINANCIALS','MIDCAPS'].some(substring => {
        const startIndex = symbol.indexOf(substring);
        if (startIndex !== -1) {
            underlying = substring;
            endIndex = startIndex + substring.length;
            expiryDate = symbol.substring(endIndex, endIndex + 7) || "Spot Price"
        }
    });

    strikePrice = symbol.substring(endIndex + 7, symbol.length - 2) || "Futures";
    optionType = symbol.substring(symbol.length - 2, symbol.length) || "--";

    const sequenceNumber = data.readBigInt64LE(34);
    const timestamp = data.readBigInt64LE(42);
    const ltp = data.readBigInt64LE(50) / BigInt(100);
    const ltpQuantity = data.readBigInt64LE(58);
    const volume = data.readBigInt64LE(66);
    const bidPrice = data.readBigInt64LE(74) / BigInt(100);
    const bidQuantity = data.readBigInt64LE(82);
    const askPrice = data.readBigInt64LE(90) / BigInt(100);
    const askQuantity = data.readBigInt64LE(98);
    const openInterest = data.readBigInt64LE(106);
    const prevClosePrice = data.readBigInt64LE(114) / BigInt(100);
    const prevOpenInterest = data.readBigInt64LE(122);

    if (optionType == "CE" || optionType == "PE") {

        // Emit the received snapshot data to the connected socket client
        socket.emit('snapshotData', {
            // streamType: "DataStream",
            key:
            {
                underlying: underlying.toString(),
                expiryDate: expiryDate.toString()
            },
            strike: strikePrice.toString(),

            optionType: optionType.toString(),
            data: {
                // sequenceNumber: sequenceNumber.toString(),
                expiryDate: expiryDate.toString(),
                timestamp: timestamp.toString(),
                ltp: ltp.toString(),
                ltpQuantity: ltpQuantity.toString(),
                volume: volume.toString(),
                bidPrice: bidPrice.toString(),
                bidQuantity: bidQuantity.toString(),
                askPrice: askPrice.toString(),
                askQuantity: askQuantity.toString(),
                openInterest: openInterest.toString(),
                prevClosePrice: prevClosePrice.toString(),
                prevOpenInterest: prevOpenInterest.toString(),
                optionType: optionType.toString()
            }

        });
    }

    if (expiryDate == "Spot Price") {
        socket.emit('snapshotData', {
            streamType:"SpotPriceUpdate",
            underlying: underlying.toString(),
            ltp: ltp.toString(),
            prevClosePrice: prevClosePrice.toString(),
            timestamp: timestamp.toString(),
        })
    }

}



// Start the server
http.listen(3000, () => {
    console.log('Server is running on port 3000');
});
